### PR TITLE
Bug Fix: String and CString casting bug (whitespaces not being ignored)

### PR DESCRIPTION
**Pull request by kushalsingh007**
- Modified the String and Cstring module to take care of casting errors for
  all the major types like int, uint, real, complex and imaginary.
- Also throws appropriate errors in case of invalid input.
- Also removing .bad and .future files for
 test/users/ferguson/cast_to_int_ignore_whitespace.chpl

### CODE REVIEW

**Comment by kushalsingh007** [comment]
@lydia-duncan : Made the PR with the needed changes in the modules section instead of runtime
Please have a look, also do let me know if I need to make some changes (for e.g str being a var ? or directory passing the modified string ?

**Comment by lydia-duncan** [comment]
Thanks!  I'm looking through this and testing now :)

File: modules/internal/CString.chpl [hunk 1]
```diff
@@ -217,21 +217,21 @@ module CString {
...
+    return c_string_to_uint8_t((x:string).strip().c_str());
   inline proc _cast(type t, x:c_string) where t == uint(16)
-    return c_string_to_uint16_t(x);
+    return c_string_to_uint16_t((x:string).strip().c_str());
   inline proc _cast(type t, x:c_string) where t == uint(32)
-    return c_string_to_uint32_t(x);
+    return c_string_to_uint32_t((x:string).strip().c_str());
   inline proc _cast(type t, x:c_string) where t == uint(64)
-    return c_string_to_uint64_t(x);
+    return c_string_to_uint64_t((x:string).strip().c_str());
```
Comment (lydia-duncan) [inline modules/internal/CString.chpl]: I think we don't have a test case for the uint cases yet, can we add that to the other pull request?

Comment (lydia-duncan) [inline modules/internal/CString.chpl]: Extra credit if you cover all the other sizes of the types being tested.  You don't have to, though :)

File: modules/internal/StringCasts.chpl [hunk 2]
```diff
@@ -71,24 +72,25 @@ module StringCasts {
 
   inline proc _cast(type t, x: string) where isIntegralType(t) {
     //TODO: switch to using qio's readf somehow
+    var str = x.strip();
     pragma "insert line file info"
-    extern proc c_string_to_int8_t  (x:c_string) : int(8);
+    extern proc c_string_to_int8_t  (str:c_string) : int(8);
```
Comment (lydia-duncan) [inline modules/internal/StringCasts.chpl]: I don't think the names need to be updated.  These aren't calls, they're extern procedure declarations.

File: modules/internal/StringCasts.chpl [hunk 3]
```diff
@@ -139,12 +141,13 @@ module StringCasts {
   }
 
   inline proc _cast(type t, x: string) where isRealType(t) {
+    var str = x.strip();
     pragma "insert line file info"
-    extern proc c_string_to_real32(x: c_string) : real(32);
+    extern proc c_string_to_real32(str: c_string) : real(32);
     pragma "insert line file info"
-    extern proc c_string_to_real64(x: c_string) : real(64);
+    extern proc c_string_to_real64(str: c_string) : real(64);
```
Comment (lydia-duncan) [inline modules/internal/StringCasts.chpl]: Same comment here

Comment (lydia-duncan) [inline modules/internal/StringCasts.chpl]: As well as in the imag and complex cases

**Comment by lydia-duncan** [comment]
Other than the comments, I think this is good.  What were you thinking about on the questions you posed?  I don't think the string being var instead of param should affect anything (if that's what you're asking), and I'm not sure what you mean about the directory question.

**Comment by lydia-duncan** [comment]
Passed std/ testing

**Comment by lydia-duncan** [comment]
Passed gasnet testing

**Comment by kushalsingh007** [comment]
@lydia-duncan : Done :smile_cat: .. Actually your idea of covering all sizes of other types was a great idea !
I found another bug in the process (which I fixed along here)
Currently only 
``` 
var a = "true";
var b = a:bool;
writeln(b);
```

was the only thing possible . Cases like 

``` 
var a = "true";
var b = a:bool(some_size_val);
writeln(b);
```
Gave away errors
So,made a minor change to fix that too :)

File: modules/internal/CString.chpl (hunk 1 above)
Comment (kushalsingh007) [inline modules/internal/CString.chpl]: Bonus part :smiley_cat: :sparkles: : Done !

**Comment by lydia-duncan** [comment]
Awesome, thanks!  I'm glad we checked :)  I'll merge this and the test PR today.

**Pull request merged**